framework/AGGRenderer/AGGRenderer/AGGRenderer.swift

```diff
@@ -21,99 +24,227 @@ public class AGGRenderer: Renderer{
     public init(width w: Float = 1000, height h: Float = 660) {
         initialized = false
         plotDimensions = PlotDimensions(frameWidth: w, frameHeight: h)
-        agg_object = initializePlot(plotDimensions.frameWidth, plotDimensions.frameHeight, plotDimensions.subWidth, plotDimensions.subHeight)
+        agg_object = initializePlot(plotDimensions.frameWidth,
+                                    plotDimensions.frameHeight,
+                                    plotDimensions.subWidth,
+                                    plotDimensions.subHeight)
         initialized = true
     }
 
-    public func drawRect(topLeftPoint p1: Point, topRightPoint p2: Point, bottomRightPoint p3: Point, bottomLeftPoint p4: Point, strokeWidth thickness: Float, strokeColor: Color = Color.black, isOriginShifted: Bool) {
+    public func drawRect(topLeftPoint p1: Pair<FloatConvertible,FloatConvertible>,
+                         topRightPoint p2: Pair<FloatConvertible,FloatConvertible>,
+                         bottomRightPoint p3: Pair<FloatConvertible,FloatConvertible>,
+                         bottomLeftPoint p4: Pair<FloatConvertible,FloatConvertible>,
+                         strokeWidth thickness: Float,
+                         strokeColor: Color = Color.black,
+                         isOriginShifted: Bool) {
         var x = [Float]()
         var y = [Float]()
-        x.append(p1.x + xOffset)
-        x.append(p2.x + xOffset)
-        x.append(p3.x + xOffset)
-        x.append(p4.x + xOffset)
-        y.append(p1.y + yOffset)
-        y.append(p2.y + yOffset)
-        y.append(p3.y + yOffset)
-        y.append(p4.y + yOffset)
-        draw_rect(x, y, thickness, strokeColor.r, strokeColor.g, strokeColor.b, strokeColor.a, isOriginShifted, agg_object)
+        x.append(Float(p1.x) + xOffset)
+        x.append(Float(p2.x) + xOffset)
+        x.append(Float(p3.x) + xOffset)
+        x.append(Float(p4.x) + xOffset)
+        y.append(Float(p1.y) + yOffset)
+        y.append(Float(p2.y) + yOffset)
+        y.append(Float(p3.y) + yOffset)
+        y.append(Float(p4.y) + yOffset)
+        draw_rect(x,
+                  y,
+                  thickness,
+                  strokeColor.r,
+                  strokeColor.g,
+                  strokeColor.b,
+                  strokeColor.a,
+                  isOriginShifted,
+                  agg_object)
     }
 
-    public func drawSolidRect(topLeftPoint p1: Point, topRightPoint p2: Point, bottomRightPoint p3: Point, bottomLeftPoint p4: Point, fillColor: Color = Color.white, hatchPattern: BarGraphSeriesOptions.Hatching, isOriginShifted: Bool) {
+    public func drawSolidRect(topLeftPoint p1: Pair<FloatConvertible,FloatConvertible>,
```

Comment:
While there's utility in making data points in a series generic, I don't know that there's as much to be gained by having things like plot borders be generic. Would there ever be cases where people would need precision beyond Float or need to plot values that would overflow standard types?

I wonder if this could be made easier to work with by having Point be a special case of Pair with two float values. Point still makes sense as a name for a type that defines physical points (upper-left, upper-right coordinates, etc.).

Reply:
Yes, we wont need precision beyond Float for drawing. I am using Pair<FloatConvertible,FloatConvertible> everywhere in all the plots. Even while calculations are being done I'm just temporarily casting them to Float. This is why I've used FloatConvertible here.